### PR TITLE
cmake: add support for Intel Compiler

### PR DESCRIPTION
**Pull request by madebr**
- Add `SDL_rdtsc` to `SDL_intrin.h`.
- Detect `_rdtsc` in cmake script
- Add CMake supports for both old and new Intel compiler (the new one is LLVM based)
- disable a few warnings emitted by (old) Intel C compiler
- fix `-Wtautological-constant-compare` warnings emitted by (new) Intel LLVM compiler.
- Test both old and new Intel compiler on CI

These are the warnings that the old Intel C Compiler warned us about:
```
/sdl/src/libm/e_exp.c(126): warning #264: floating-point value does not fit in   required floating-point type
  		if(x > o_threshold) return huge*huge; /* overflow */
  		                           ^
/sdl/src/libm/e_exp.c(131): warning #239: floating point underflow
  	    if(x < u_threshold) return twom1000*twom1000; /* underflow */
  	                               ^
 /sdl/src/libm/e_log.c(98): warning #39: division by zero
  	    if (hx<0) return (x-x)/zero;	/* log(-#) = NaN */
  	                           ^
/sdl/src/libm/e_log10.c(74): warning #39: division by zero
              if (hx<0) return (x-x)/zero;        /* log(-#) = NaN */
                                     ^
/sdl/src/video/SDL_blit_copy.c(150): (col. 9) warning #13203: No EMMS instruction before call to function
```
Because the libm code is 3rd party code, I chose to ignore those specific warnings.

About the EMMS warning, I'm not really sure what's that about.
[MMX require a call to `EMMS` after the last mmx instruction](https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-8/the-emms-instruction-why-you-need-it.html), [which the code does](https://github.com/libsdl-org/SDL/blob/0ef85873a2b15a3f170d09deba6989afa145ec43/src/video/SDL_blit_copy.c#L92) (`emms` == `_mm_empty`).
So I disabled that warning too.

The only warning of the new oneAPI/LLVM Intel compiler was a `-Wtautological-constant-compare` warning.
I fixed this one by using [`fpclassify`](https://man7.org/linux/man-pages/man3/fpclassify.3.html) instead.

### CODE REVIEW

File: include/SDL3/SDL_intrin.h [hunk 1]
```diff
@@ -108,13 +110,14 @@ _m_prefetch(void *__P)
 #endif
 
 #if defined(__x86_64__) || defined(_M_X64) || defined(__i386__) || defined(_M_IX86)
-# if ((defined(__GNUC__) && (__GNUC__ >= 10)) || (defined(__clang__) && __clang_major__ >= 4) || defined(_MSC_VER)) && !defined(SDL_DISABLE_RDTSC)
+# if (defined(_MSC_VER) || (defined(__GNUC__) && (__GNUC__ > 4 || __GNUC__ == 4 && __GNUC_MINOR__ >= 4)) || defined(__clang__) || defined(__ICC)) && !defined(SDL_DISABLE_RDTSC)
```
[the remark this answers is not in the document]
Comment (madebr) [reply include/SDL3/SDL_intrin.h:113]: For symmetry with the `SDL_HasXXX`  SDL functions.

[the remark this answers is not in the document]
Comment (madebr) [reply include/SDL3/SDL_intrin.h:113]: Aight, I'll move it to the test.

[the remark this answers is not in the document]
Comment (madebr) [reply include/SDL3/SDL_intrin.h:113]: Done in 705aa7567265318dd8a56e865516d425688dfcc3

[the remark this answers is not in the document]
Comment (slouken) [reply include/SDL3/SDL_intrin.h]: These days we should pull RDTSC entirely and rely on the OS time calls, as they properly handle multi-core synchronization and so forth.

[the remark this answers is not in the document]
Comment (madebr) [reply include/SDL3/SDL_intrin.h]: So okay to remove `SDL_HasRDTSC` from SDL3?

**Comment by ccawley2011** [comment]
> About the EMMS warning, I'm not really sure what's that about.
[MMX require a call to EMMS after the last mmx instruction](https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-8/the-emms-instruction-why-you-need-it.html), [which the code does](https://github.com/libsdl-org/SDL/blob/0ef85873a2b15a3f170d09deba6989afa145ec43/src/video/SDL_blit_copy.c#L92) (emms == _mm_empty).
So I disabled that warning too.

It's probably complaining about the call to `SDL_memcpy()` - since it's not inlined, the compiler can't tell if it'll attempt any floating point operations.

File: include/SDL3/SDL_intrin.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (slouken) [reply include/SDL3/SDL_intrin.h]: Yes please.

[the remark this answers is not in the document]
Comment (madebr) [reply include/SDL3/SDL_intrin.h]: Removed in c0ee58025b1ea20af8ee703d765e2024a99c74c5

**Comment by madebr** [comment]
> It's probably complaining about the call to `SDL_memcpy()` - since it's not inlined, the compiler can't tell if it'll attempt any floating point operations.

You're right. What do you suggest?
Add a `_mm_empty()` before `SDL_memcpy`?
Or replace it with `while (remain--) { *dst++ = *src++; }`

**Comment by slouken** [comment]
Either of those is probably fine. Note that system implementations of memcpy() may actually use SIMD, so that's a legitimate warning.

**Comment by madebr** [comment]
I went with the loop because the emms instruction has some cost, whereas the loop can be optimized.

**Comment by sezero** [comment]
Can we also apply the necessary parts of this to sdl2-compat and possibly to sdl12-compat?

**Pull request merged**